disco_aws_automation/disco_config.py

```diff
@@ -59,3 +60,72 @@ def open_normalized(*path_components, **kwargs):
     return open(real_path, **kwargs)
+
+
+class AsiaqConfig(ConfigParser):
+    """
+    Wrap configuration-reading shortcuts around the ConfigParser object.
+
+    All methods accept an "environment" parameter, but recommended usage is to pass in the
+    desired environment name at construction time, rather than at call time.
+    """
+
+    S3_BUCKET_BASE_OPTION = 's3_bucket_base'
+    S3_BUCKET_SUFFIX_OPTION = 's3_bucket_suffix'
+    DEFAULT_ENVIRONMENT_OPTION = 'default_environment'
+
+    def __init__(self, environment=None):
+        ConfigParser.__init__(self)  # ConfigParser is an old-style class
+        self._real_environment = environment
+
+    @property
+    def environment(self):
+        """
+        The environment name passed in at construction, or the default value for this config file.
+
+        Lazily populated since the config file has not yet been read at initialization time.
+        """
+        if not self._real_environment:
+            self._real_environment = self.get(DEFAULT_CONFIG_SECTION, self.DEFAULT_ENVIRONMENT_OPTION)
+        return self._real_environment
+
+    def get_asiaq_option(self, option, section=DEFAULT_CONFIG_SECTION, environment=None,
+                         required=True, default=None):
+        """
+        Get a value from the config, checking first for an environment-specific value, then
+        a generic value, then an env-specific default value in the default section, then a
+        non-env-specific default value in the default section.
+
+        In the case where none of these options has been set, if the "required" option is False,
+        return the value of the "default" option; otherwise, raise NoOptionError.
+        """
+        if required and (default is not None):
+            raise ProgrammerError("Using the 'default' option when 'required' is True makes no sense.")
+        if not environment:
+            environment = self.environment
+        env_option = "{0}@{1}".format(option, environment)
+        default_option = "default_{0}".format(option)
+        default_env_option = "default_{0}".format(env_option)
+
+        if self.has_option(section, env_option):
+            return self.get(section, env_option)
+        if self.has_option(section, option):
+            return self.get(section, option)
+        elif self.has_option(DEFAULT_CONFIG_SECTION, default_env_option):
+            return self.get(DEFAULT_CONFIG_SECTION, default_env_option)
+        elif self.has_option(DEFAULT_CONFIG_SECTION, default_option):
+            return self.get(DEFAULT_CONFIG_SECTION, default_option)
+
+        if required:
+            raise NoOptionError(option, section)
+        return default
+
+    def get_asiaq_s3_bucket_name(self, bucket_tag, environment=None, separator='--'):
+        """Construct a standardized bucket name based on configured prefix and suffix values."""
+        bucket_base = self.get_asiaq_option(self.S3_BUCKET_BASE_OPTION, environment=environment)
```

Comment:
Need to add extra options to the file, yes.  Handling backward compatibility by not actually using it anywhere yet, but assuming there will be a brief transition period when both configurations are available in disco_aws.ini.  The only thing this is currently slated to displace is the SSM bucket configuration, which is what I stole the pattern from—I'm expecting to also use it for data pipeline backups, but that's new code that hasn't been pushed yet, so it has no backward-compatibility issue.

The general idea is to reduce the number of places we need to modify configuration files to introduce a new predictable-bucket S3 stash, but in the short term it does increase configuration changes.

There's no reason _not_ to use this class with files other than disco_aws.ini, though it's not per se all that useful unless you're using environment-based variable changes: since it inherits from ConfigParser, you can go ahead and ignore that it's not a vanilla ConfigParser for many cases.